Documentation of the generic-server (#93)
adorsys/didcomm-mediator-rs: generic-server/README.md

# generic-server

This server provides a framework to build versatile applications by aggregating functionalities offered by configurable plugins.

> 

Author: @Hermann-Core  Please review this again.

# generic-server

This server provides a system for building versatile applications by integrating functionalities offered by configurable plugins.
## Features

* **Plugin System:** Enables developers to extend server behavior through custom plugins.
* **Routing and Transport:** Handles communication between clients and plugins using industry-standard protocols (e.g., HTTP).

# Usage
## Extending Server Functionalities with Plugins
### To add a new plugin:

**1. Implement the `Plugin` Trait:**

 Define the necessary methods (e.g., name(), mount(), and routes()).

**2. Register the Plugin:**

>  Add the plugin to the static PLUGINS array.This array are located in your main server configuration file or a central location where the server initializes and manages plugins.

Reviewer: ```suggestion
 Add the plugin to the static PLUGINS array (located in `src/plugin/mod.rs`).
```

**3. Utilize the Plugin:**

 The server will automatically handle routing based on the newly added plugin's routes.

## Example Plugin Implementation
```rust
struct MyPlugin;

impl Plugin for MyPlugin {
    fn name(&self) -> &str {
        "MyPlugin"
    }

    fn mount(&self) -> Result<(), PluginError> {
        // Initialization logic here
        Ok(())
    }
    fn unmount(&self) -> Result<(), PluginError> {
        // Initialization logic here
        Ok(())
    }
    fn routes(&self) -> Router {
        // Define and return routes here
        Router::new().route("/myplugin", get(my_plugin_handler))
    }
}

> ```

Reviewer 2: Maybe you can also add another example here showing how the plugin can be added to the static `PLUGINS` array ?

# generic-server

This server provides a system for building versatile applications by integrating functionalities offered by configurable plugins.
## Features

* **Plugin System:** Enables developers to extend server behavior through custom plugins.
* **Routing and Transport:** Handles communication between clients and plugins using industry-standard protocols (e.g., HTTP).

# Usage
## Extending Server Functionalities with Plugins
### To add a new plugin:

> 
> **1- Implement the `Plugin` Trait:**
> 
> - Define the necessary methods (e.g., name(), mount(), and routes()).
> 
> **2-Register the Plugin:**
> 
> - Add the plugin to the static PLUGINS array.

Reviewer: Improve the styling:
- Use conventional markdown numbered lists.
- There is no need to define a list when you only have one item. For example, I'm using a list to make suggestions here 'cause I have at least two of them.
- Preview your markdown code. You can use a live editor such as https://dillinger.io/.
<br>

```suggestion

**1. Implement the `Plugin` Trait:**

Define the necessary methods (e.g., name(), mount(), and routes()).

**2. Register the Plugin:**

Add the plugin to the static PLUGINS array.
```

> - Add the plugin to the static PLUGINS array.

Reviewer: Indicate where to find that array.

**3-Load the Plugin:**

> - Ensure the plugin container is re-initialized to include the new plugin.**

Reviewer: This step is not necessary, the plugins are loaded automatically as the server boots. I can't recall the plugin container being able to be re-initialized.
